### PR TITLE
[ML] Reset anomaly detection job API

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
Adds a new API that allows a user to reset
an anomaly detection job.

To use the API do:

```
POST _ml/anomaly_detectors/<job_id>/_reset
```

The API removes all data associated to the job.
In particular, it deletes model state, results and stats.

However, job notifications and user annotations are not removed.

Also, the API can be called asynchronously by setting the parameter
`wait_for_completion` to `false` (defaults to `true`). When run
that way the API returns the task id for further monitoring.

In order to prevent the job from opening while it is resetting,
a new job field has been added called `block_reason`. This can
take a value from ["delete", "reset", "revert"] as all these
operations should block the job from opening. The delete action
has already been blocking the job by setting the `deleting` field.
However, in order not to introduce different booleans for each
action, `block_reason` should be the way to do this onwards.

Finally, this commit also sets the `block_reason` to `revert` when
the revert snapshot API is called as a job should not be opened
while it is reverted to a different model snapshot.

### CODE REVIEW

**Review by droberts195 — COMMENTED** [review]
Looks good.

The comment about revert during revert not being safe is probably a nightmare to do anything about, so maybe ignore it in this PR given that we didn't prevent it in the past and nobody apparently noticed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:140]: This isn't required. When writing the XContent body, `block_reason` isn't written when its `null`. So, it should never be a `null` value when parsing from XContent

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:912]: May I suggest
```suggestion
            this.deleting = (this.blockReason == BlockReason.DELETE);
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:901]: I don't know if we can do a `deleting = false` here for BWC purposes (what if we parse a `null` blockReason but `deleting=true` from XContent???). This method is only used from the XContent parser.

However, we NEVER write out a `null` value to the xcontent body, so I don't think this method is ever called anyways.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/config/JobTests.java
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/config/JobTests.java:597]: I think we should have some BWC tests for XContent parsing. 

- We should make sure that an old job with `deleting: true` when parsed has `BlockReason.DELETE`

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java:122]: 🔧

What if the master node has changed? Is it possible for the master node to change without the previous master node stopping?

This may only occur for a smaller period of time when a cluster first starts, or if there are network connection issues...So it may not be a concern.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportRevertModelSnapshotAction.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportRevertModelSnapshotAction.java:109]: Agreed, I don't think this should continue if the block reason has been set at all.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/job/RestResetJobAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/job/RestResetJobAction.java:53]: I don't 100% understand this.

Does this create a "local task" on this coordinating node that waits around until the master action is complete?

File: docs/reference/ml/anomaly-detection/apis/reset-job.asciidoc
Comment (benwtrent) [inline docs/reference/ml/anomaly-detection/apis/reset-job.asciidoc:73]: it might be good to indicate that to check on the status of the task, you shouldn't call the same API again and instead should use the `tasks` API.

**Review by benwtrent — COMMENTED** [review]
I don't see the `org.elasticsearch.xpack.security.operator.Constants` updated, so I bet that is why the build failed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:140]: When we update the job document to clear `block_reason` we do set `block_reason` to `null` explicitly.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/job/RestResetJobAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/job/RestResetJobAction.java:53]: Correct.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:140]: @dimitris-athanasiou true, but is `null` ever written to the xcontent builder and thus written to the doc? From what I can tell, it is not. It is simply not written to the doc at all. 

There is a big difference between `block_reason: null` and `block_reason` not being in the doc at all.

Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:140]: It is. See `JobConfigProvider.updateJobBlockReason`.

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:140]: Ah, we are writing a map directly and updating the doc. 

Why are we explicitly saying `null` for the value instead of removing the field?

We now have two things that indicate that it isn't blocked, the field not existing or the field being explicitly null.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/Job.java:901]: I see now that we are indeed writing `null` directly and by passing the job's ToXContent values.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportRevertModelSnapshotAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportRevertModelSnapshotAction.java:109]: Makes sense, I'll remove the `revert` clause.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java:122]: I think when it comes to best effort this is all we can do. The worse case scenario is we launch a second reset which isn't catastrophic.

**Review by szabosteve — APPROVED** [review]
Docs LGTM. Thanks for adding them. I left two minor comments, please take or leave them!

**Comment by dimitris-athanasiou** [comment]
I have reworked this in order to include the task id. The new job field now looks like:

```
...
"blocked": {
  "reason": "reset",
  "task_id": "abc:123"
}
...
```

I kept the update of the `blocked` field internal. Revert now checks if there is a revert task running and if not allows to run even if the job is blocked due to revert. This means the user can just call revert again to fix a job stuck with revert (in case the node failed half-way, etc.)

Could you please take another look?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java:367]: Now that we have the `task_id` in the `blocked` object, is this still necessary?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java:148]: I think this and the above list method can be adjusted to look at the current `blocked.task_id` and if the task doesn't exist, start another one and update the doc.

**Comment by droberts195** [comment]
One more thing I just remembered is that even though we have a shiny new auto-generated HLRC the manually created HLRC needs to be complete and supported until 7.last.  So at some point (not necessarily in this PR), please can you add this new API to the HLRC?

**Comment by dimitris-athanasiou** [comment]
@droberts195 

> One more thing I just remembered is that even though we have a shiny new auto-generated HLRC the manually created HLRC needs to be complete and supported until 7.last. So at some point (not necessarily in this PR), please can you add this new API to the HLRC?

Yes, I was planning to do so in a separate PR that will go only on the `7.x` branch after the backport.

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/bwc

**Review by benwtrent — APPROVED** [review]
I think this looks good.

It may be good to include the task_id in the error messages or something to indicate how they can check when the thing is over.

**Review by droberts195 — APPROVED** [review]
LGTM, if you could remove that piece of temporary debug before merging

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java:340]: I know that if `Blocked.Reason.RESET` then `job.getBlocked().getTaskId()` should never be null. But, if it is ever null, this would blow up. 

Same goes for any other tasks request. If we make a tasks request with a `null` taskId, this will go boom. 

I didn't see anything in this PR that struck me as the only time the taskID could be null is if the `Blocked.Reason.DELETE` and we don't ever check the task ID there. 

Just wanted to make sure we all knew this :)

File: test/framework/src/main/java/org/elasticsearch/test/AbstractWireTestCase.java
Comment (dimitris-athanasiou) [reply test/framework/src/main/java/org/elasticsearch/test/AbstractWireTestCase.java:88]: Oups! Well spotted!

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteJobAction.java:340]: Yes, exactly.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Pull request merged**
